### PR TITLE
Remove `dest_asyncwrite`, reorg codecs, consolidate `collect_ready`, fix #216

### LINKED ISSUE #216: having both sink_async and write_async is confusing for users
are they meaningfully different? can we make it easier to understand? Else document better.

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow/src/util/mod.rs [hunk 1]
```diff
@@ -98,11 +97,22 @@ where
 }
 
 /// Receives available items in an `UnboundedReceiverStream` into a `FromIterator` collection.
-pub fn recv_into<C, T>(recv: &mut UnboundedReceiverStream<T>) -> C
+pub fn recv_into<C, S>(stream: S) -> C
```
Comment (jhellerstein) [inline hydroflow/src/util/mod.rs:100]: What's this change for?

File: hydroflow/tests/surface_async.rs [hunk 2]
```diff
@@ -263,3 +264,106 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
 
     Ok(())
 }
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_sink_bounded_channel() {
+    // This is for testing. Don't actually use unbounded channels, as their input will end up
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:270]: Comment is a bit cryptic. If I understand right you mean:
"In ordinary code we would use unbounded channels, but that removes our ability to control buffering in a test. For this test, we restrict ourselves to a bounded channel (size 5)." I also don't understand what it says about `send.send(x).unwrap())`... is that a vestigial comment?

File: hydroflow/tests/surface_async.rs [hunk 3]
```diff
@@ -263,3 +264,106 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
+async fn asynctest_dest_sink_duplex() {
+    use bytes::Bytes;
+    use tokio::io::AsyncReadExt;
+    use tokio_util::codec;
+
+    // Like a channel, but for a stream of bytes instead of discrete objects.
+    let (asyncwrite, mut asyncread) = tokio::io::duplex(256);
+    // Now instead handle discrete byte lists by length-encoding them.
+    let mut sink = codec::LengthDelimitedCodec::builder()
+        .length_field_length(1)
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:299]: Is `length_field_length(1)` recommended design, or an extreme case you're using for testing? Where would you document that?

File: hydroflow/tests/surface_async.rs [hunk 4]
```diff
@@ -263,3 +264,106 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
+    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
+        .await
+        .expect_err("Expected time out");
+
+    // Only 5 elemts received due to buffer size
+    let out: Vec<_> = recv_into(&mut recv);
+    assert_eq!(&[0, 1, 2, 3, 4], &*out);
+}
+
+#[tokio::test(flavor = "current_thread")]
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:289]: Now I'm real confused. I thought `dest_asyncwrite` is for byte streams and `dest_sink` is for collections of discrete objects.  Which one are you testing here?

File: hydroflow/tests/surface_async.rs [hunk 5]
```diff
@@ -263,3 +264,106 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
+        ]) -> dest_asyncwrite(asyncwrite);
+    };
+    tokio::time::timeout(std::time::Duration::from_secs(1), flow.run_async())
+        .await
+        .expect_err("Expected time out");
+
+    let mut buf = Vec::<u8>::new();
+    asyncread.read_buf(&mut buf).await.unwrap();
+    // `\x05` is length prefix of "5".
+    assert_eq!(b"helloworld", &*buf);
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:339]: This is helpfully distinct from the previous case.

File: hydroflow_lang/src/graph/ops/dest_asyncwrite.rs [hunk 6]
```diff
@@ -13,6 +13,32 @@ use quote::quote_spanned;
...
 ///
 /// Note this operator must be used within a Tokio runtime.
+///
+/// ```rustbook
+/// #[tokio::test]
+/// async fn test_dest_asyncwrite() {
+///     use tokio::io::AsyncReadExt;
+///
+///     // Like a channel, but for a stream of bytes instead of discrete objects.
+///     // This could be an output file, network port, stdout, etc.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/dest_asyncwrite.rs:23]: That is, every run of bytes that you send into this `dest` is appended to the previously-sent bytes, without delimiters.

File: hydroflow_lang/src/graph/ops/dest_sink.rs [hunk 7]
```diff
@@ -12,6 +12,37 @@ use quote::quote_spanned;
...
 /// Note this operator must be used within a Tokio runtime.
+///
+/// ```rustbook
+/// #[tokio::test]
+/// async fn test_dest_sink() {
+///     use bytes::Bytes;
+///     use tokio::io::AsyncReadExt;
+///     use tokio_util::codec;
+///
+///     // Like a channel, but for a stream of bytes instead of discrete objects.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/dest_sink.rs:23]: Yeah I'm still confused.

**Review by jhellerstein — COMMENTED** [review]
I'm now understand `dest_asyncwrite` but am confused by your description of `dest_sync`

File: hydroflow/src/util/mod.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow/src/util/mod.rs:28]: Makes this method more general

File: hydroflow/tests/surface_async.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow/tests/surface_async.rs:295]: Just for testing so we dont have to worry about endianness

File: hydroflow/tests/surface_async.rs [hunk 8]
```diff
@@ -258,8 +255,113 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
-    let seen: Vec<_> = hydroflow::util::collect_ready(seen_recv).await;
+    let seen: Vec<_> = collect_ready(seen_recv);
     rassert_eq!(&std::array::from_fn::<_, MAX, _>(|i| i), &*seen)?;
 
     Ok(())
 }
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_sink_bounded_channel() {
+    // In this example we use an _bounded_ channel for our `Sink`. This is for demonstration only,
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:266]: typo "an" --> "a"

File: hydroflow/tests/surface_async.rs [hunk 9]
```diff
@@ -258,8 +255,113 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
+        .await
+        .expect_err("Expected time out");
+
+    // Only 5 elemts received due to buffer size
+    let out: Vec<_> = collect_ready(&mut recv);
+    assert_eq!(&[0, 1, 2, 3, 4], &*out);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_sink_duplex() {
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:286]: this is complicated and I'm not sure if you intend for users to do this?

File: hydroflow/tests/surface_async.rs [hunk 10]
```diff
@@ -258,8 +255,113 @@ pub async fn test_futures_stream_sink() -> Result<(), Box<dyn Error>> {
...
+        .expect_err("Expected time out");
+
+    let mut buf = Vec::<u8>::new();
+    asyncread.read_buf(&mut buf).await.unwrap();
+    // `\x05` is length prefix of "5".
+    assert_eq!(b"\x05hello\x05world", &*buf);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn asynctest_dest_asyncwrite_duplex() {
```
Comment (jhellerstein) [inline hydroflow/tests/surface_async.rs:316]: similarly complicated. I get uncomfortable when I see calls to raw `tokio::io` and codecs, to be honest.

File: hydroflow/tests/surface_codegen.rs [hunk 11]
```diff
@@ -2,8 +2,7 @@ use std::collections::{HashMap, HashSet};
```
Comment (jhellerstein) [inline hydroflow/tests/surface_codegen.rs:2]: why these changes? is `recv_into` deprecated or are you trying to demonstrate something different, or...?

File: hydroflow_lang/src/graph/ops/dest_sink.rs [hunk 12]
```diff
@@ -7,11 +7,69 @@ use quote::quote_spanned;
...
+/// A `Sink` is a thing into which values can be sent, asynchronously. For example, sending items
+/// into a bounded channel.
 ///
 /// Note this operator must be used within a Tokio runtime.
+///
+/// ```rustbook
+/// # #[tokio::main]
+/// # async fn main() {
+/// // In this example we use an _bounded_ channel for our `Sink`. This is for demonstration only,
+/// // instead you should use [`hydroflow::util::unbounded_channel`]. A bounded channel results in
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/dest_sink.rs:19]: what's the benefit of the bounded channel in documentation?

**Review by jhellerstein — APPROVED** [review]
Not a deep read. I have some questions on the low-level nature of the API surface in the examples. Maybe we just need to separately document the easy cases from the sophisticated ones, and justify the sophisticated ones.

Approve to merge as you see fit.

File: hydroflow/tests/surface_codegen.rs (hunk 11 above)
Comment (MingweiSamuel) [reply hydroflow/tests/surface_codegen.rs:2]: Originally there were two methods, `recv_into` and `collect_ready`, both of which collecting ready items from an async stream and put them into a collection. However `recv_into` only worked on channel recv streams. Meanwhile `collect_ready` was async even though it didn't really need to be. So they are now both combined into one `collect_ready` which is synchronous and handles all stream types

File: hydroflow_lang/src/graph/ops/dest_sink.rs (hunk 12 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/ops/dest_sink.rs:19]: Added
```
(We can't use
/// // unbounded here since unbounded channels are synchonous to write to and therefore not
/// // `Sink`s.)
```

File: hydroflow/tests/surface_async.rs (hunk 9 above)
Comment (MingweiSamuel) [reply hydroflow/tests/surface_async.rs:286]: not necessarily intended for users, but works as a test
